app/channel.go

```diff
@@ -330,6 +330,17 @@ func (a *App) GetOrCreateDirectChannel(userID, otherUserID string, channelOption
 		return channel, nil
 	}
 
+	if *a.Config().TeamSettings.RestrictDirectMessage == model.DIRECT_MESSAGE_TEAM &&
+		!a.SessionHasPermissionTo(*a.Session(), model.PERMISSION_MANAGE_SYSTEM) {
+		commonTeamIDs, err := a.GetCommonTeamIDsForTwoUsers(userID, otherUserID)
+		if err != nil {
+			return nil, err
+		}
+		if len(commonTeamIDs) == 0 {
+			return nil, model.NewAppError("createDirectChannel", "api.channel.create_channel.direct_channel.team_restricted_error", nil, "", http.StatusForbidden)
+		}
+	}
```

Comment:
I'm not sure why we'd make an exception for the `PERMISSION_MANAGE_SYSTEM` permission here given that this setting is configured in the settings. If we want it to be part of the permissions system we could create a permission `message_non_intersecting_team_members`.

Reply:
The idea was that sysadmins could override the direct message policy, i.e. sysadmins can create direct channels between two arbitrary users. Or is this not something we want?

Reply:
The need for that functionality would be a PM question, but in terms of implementation: having the logic partially in the config and partial in the permission system seems like an exception that we should use sparingly.

Reply:
https://mattermost.atlassian.net/browse/MM-7968?focusedCommentId=110111